### PR TITLE
feat: Complete test suite migration to RSpec + Cucumber

### LINKED ISSUE #964: Fix the neighbourhood partner policy testing suite
@alexandria-gfsc and I bashed our head against this for a bit and can't figure out what the issue is.

It seems to work on the live site but mysteriously fails to go green. The testing code is currently too hard to reason about, this could be to do with postcodes, factory bot, something IRL, something we overlooked. This is too much. We had to skip some tests.

This prob needs stripping back and reevaluating at a later date after a break from it.

### LINKED ISSUE #1478: Build a Normal Island fake neighbourhood for development
## User story

As an owner of PlaceCal I need to be able to roll PlaceCal out anywhere in the world so that I can make everyone's lives better.

As a developer of PlaceCal I need to not need to import the UK's geography and live data to work on the service so that I can work quickly within a smaller context.

## Acceptance criteria

- [ ] Local development happens on Normal Island, a fake geography with a fake geographic hierarchy
- [ ] A set of seed files I can understand by reading then in one sitting has everything I need to be able to have a blank slate geography for Normal Island to work with
- [ ] We seed at least one of every entity type, too
- [ ] We have tests to confirm the service behaves correctly for UK data
- [ ] The service still works with the UK's geography on staging and production
- [ ] There is an illustration of Normal Island we can refer to clearly for development
- [ ] The makefile and readme have been updated to use seed files and not prod db copy

## Follow up work

We then want to seed more interesting partners and all the rest to give us a "lived in" development environment.

## Implementation notes & questions

A side effect of this (and secretly the goal of it) is that we completely decouple the UK's specifics from PlaceCal, so PlaceCal works seamlessly with any geography.

## Implementation plan

(derived from #1947)

- [ ] clear out old seeding code
  - db/seeds and any migration that populates / manipulates data should be blanked
- [ ] generate actual seed data (may happen concurrently with step below)
- [ ] write seeding script
  - [ ] should be easy to use
  - [ ] should reset database completely
- [ ] update documentation

### LINKED ISSUE #1919: Test suite review / fixes
## Description

The test suite has a number of problems that, whilst not impacting work directly, can cause real problems when dealing with entirely separate issues. This results in lost productivity as we are fixing (or working around) these problems rather than delivering utility to our end-users.

- messy code in test_helper
    > move it into helper modules

- a bunch of empty test files
    > delete these

- postcodes, neighbourhoods are related through the geocoder and VCR stubs but this is basically entirely invisible
    > define some kind of helper to generate valid postcode <-> neighbourhood pairs

- fixtures
    > remove them, fix broken tests, don't use them in future (only factorybot factories)

- neighbourhoods are VERY sensitive to duplicates
    > not really a test suite problem but neighbourhoods SHOULD be unique.

- tests in controller should be in integration and tests in integration should be in controller
    > review and tidy up

- inconsistent code style
    > some tests are written in the minitest DSL and some are written in the subclass method override style. pick one

- are our factories trying to be too helpful? they were written a long time ago when the behaviour was simpler to construct but new behaviours are valid and useful. so in stead we have to tap dance around the factories building our own models, which is brittle.
    > review factories. follow the policy of least surprise: don't try to be clever in the factory

This is not exhaustive and I will add more if I find them

### LINKED ISSUE #2073: Remove references to specific partners and partnerships from codebase
## Description

There are various references to IRL partners in our database.

We should prob remove these before doing a public launch.

Locations include [iCal regexes](https://github.com/geeksforsocialchange/PlaceCal/pull/2068), tests, and some hard coded routes.

## Motivation

Our codebase should not reference actual live groups and instead be a set of data managed by us.

## Acceptance Criteria

- [x] #2409
- [ ] Rest of it to be addressed in future

### LINKED ISSUE #2263: [Bug]: rails-controller-testing gem looks like it is no longer maintained
## Description

This gem is used to allow rails 5 testing functionality into rails 6. As we're now on 7 maybe it's time to refactor these tests to move away from old syntax and then we can remove this gem?

https://github.com/rails/rails-controller-testing

This came out of https://github.com/geeksforsocialchange/PlaceCal/issues/2159

### LINKED ISSUE #2460: Test suite is poorly organised
## Description

From the perspective of coming to the code relatively cold:

- Tests seem to be in no particular order
- Tests use a wide range of formatting
- File names don't always match files
- Some test files are far too long
- There are not clear and easy to find tests covering common actions e.g. creating a new partner
- Test names are often hard to understand, in broken english, and are using a wide variety of tenses
- Personal opinion: having single quotes around test names is a pain as you have to redo the quotes if you want to use a ' symbol which just adds to cognative overload for no real benefit. This is a Rubocop thing and I would prefer test names are double quoted for easier readibility.

## Motivation

It's hard to get an idea what PlaceCal is meant to do from reading the tests

### LINKED ISSUE #2557: Specific importer tests are poorly scoped
## Description

Bug fixing things to do with the importer is tricky as there's a lot of tests that look a lot to me like magic number assertations. For example, this test asserts that there should be 15 events - why?

https://github.com/geeksforsocialchange/PlaceCal/blob/97c5c2570fd27150e63318a24172db9783e8fa96/test/jobs/calendar_importer/calendar_importer_task_test.rb#L132

An example of why this is confusing is that if I remove the `!` here, the event count changes from 17 to 15.

https://github.com/geeksforsocialchange/PlaceCal/blob/4ea8af5a01b552417f7c708fdb60440a45377860/app/jobs/calendar_importer/parsers/ld_json.rb#L116

I think overall we need smaller and more specific unit tests for each stage of the process rather than these monolithic end to end ones.

## Motivation

Importer logic is hard to reason about and test values seem arbritrary. This is a good recipe for introducing a lot of magic number bugs, and doesn't help us understand more about the code.

### LINKED ISSUE #2573: Add regression tests for existing frontend functionality
## User story

As a developer or project owner, I need to understand what the frontend is meant to do, in order that I can reason about changes to it and prevent things that work now from breaking.

## Acceptance criteria

- [ ] There is a BDD test suite that describes the desirable functionality
- [ ] Events have their behaviour described in tests
- [ ] Partners have their behaviour described in tests
- [ ] Sites have their behaviours described in tests
- [ ] Project README is updated to reflect the testing idiom

## Implementation notes & questions

Proposal is to set up a [Spinach](https://github.com/codegram/spinach) test suite that describes the functions. @pedanticgit's https://github.com/fishpercolator/name.pn/ repo has an example how to do this,

## Implementation plan

_To be written by the developer_

### LINKED ISSUE #2591: Capybara configuration is very unhappy
## Description

There's a ton of workarounds in our Capybara config and they keep spiralling. There's config in about 4 different places. We shouldn't need all these wait functions, and the console output should be clean.

The latest issue is that the first `visit` command always fails with below error. The setup block currently adds in a visit around everything but we shouldn't need this.

Probably the best thing to do is write something from scrach in `cuprite`.

```
2022-11-08 00:00:00 +0000 HTTP parse error, malformed request: #<Puma::HttpParserError: Invalid HTTP format, parsing fails. Are you trying to open an SSL connection to a non-SSL Puma?>
```

## Motivation

Tests passing reliably is a neccesity for a happy codebase.

### LINKED ISSUE #2704: Flaky tests are getting more unreliable
## Description

We have had an occasionally flaky test go from what feels like a 1/10 failure rate to a 4/10 and it's getting annoying. They look like this.

Probably we want to think about replacing this rather than doing too much work to make it work again. #2573 maybe.

```
 Done in 0.25s.
Running 17 tests in a single process (parallelization threshold is 50)

# Running tests with run options --seed 12317:

.......[Screenshot Image]: /home/runner/work/PlaceCal/PlaceCal/tmp/capybara/failures_test_visiting_the_index.png
E.........

Finished tests in 105.388306s, 0.1613 tests/s, 1.3759 assertions/s.


Error:
CreateAdminUsersTest#test_visiting_the_index:
Capybara::ElementNotFound: Unable to find field "New password" that is not disabled
    test/system/create_admin_users_test.rb:55:in `block in <class:CreateAdminUsersTest>'

17 tests, 145 assertions, 0 failures, 1 errors, 0 skips

== Command ["bin/rails test:system"] failed ==
```

## Motivation

Why is it important?

### PR DESCRIPTION
**Pull request by kimadactyl**
## Summary

Complete migration of the PlaceCal test suite from Minitest to RSpec + Cucumber with Normal Island fictional test data.

## Implementation Phases

| Phase | Description | Status |
|-------|-------------|--------|
| 1 | Foundation Setup (gems, config, move legacy tests) | ✅ Complete |
| 2 | Normal Island Data & Factories | ✅ Complete |
| 3 | Model Specs | ✅ Complete |
| 4 | Component Specs | ✅ Complete |
| 5 | Policy Specs | ✅ Complete |
| 6 | Request Specs | ✅ Complete |
| 7 | Job Specs (Calendar Importer) | ✅ Complete |
| 8 | System Specs | ✅ Complete |
| 9 | Cucumber Features | ✅ Complete |
| 10 | Development Seeds & Cleanup | ✅ Complete |

### Phases 1-2: Foundation & Normal Island
- Configure RSpec, Cucumber, FactoryBot, VCR, DatabaseCleaner
- Create `lib/normal_island.rb` with fictional geography (NO country code)
- Build comprehensive factory set for all models using Normal Island data
- Update development seeds to use Normal Island

### Phases 3-7: Core Specs
- **Model Specs** (14 files): Partner, Event, Calendar, User, Site, Tag, Neighbourhood, Address, Article, Collection, OnlineAddress, CalendarState, and scope/validation specs
- **Component Specs** (4 files): Address, Breadcrumb, PartnerFilter, PartnerPreview components
- **Policy Specs** (4 files): Partner, User, Article, Calendar policies
- **Request Specs** (18 files): GraphQL queries (events, partners, sites, articles), Admin controllers (partners, sites, articles, home, neighbourhoods, calendars, users), Public pages (events, partners, pages, robots, collections, articles, sites)
- **Job Specs** (13 files): Calendar importer parsers and strategies
- **Helper Specs** (5 files): Mailer, Articles, Partners, Users, Calendars helpers
- **Mailer Specs** (2 files): JoinMailer, Devise invitation mailer

### Phase 8: System Specs
- 10 system spec files for admin UI and authentication testing
- Select2 dropdown helpers for JavaScript interactions
- Shared admin login context
- GraphQL API network tests
- **Flaky test fixes**: Dynamic ActionMailer port configuration, Capybara session reset between tests

### Phase 9: Cucumber Features
- **11 feature files with 23 BDD scenarios** (all passing):
  - Admin: Sites (4), Articles (5), Tags (5), Calendars (2), Users (2), Partners (3)
  - Public: News (4), Site Browsing (5), Browse Events (3), Browse Partners (2)
  - Authentication (3)
- 8 step definition files (authentication, navigation, partners, events, calendars, articles, sites, tags)
- Configured Cucumber environment with FactoryBot, Capybara, and subdomain routing
- All features use `@javascript` tag for Chrome headless browser testing

### Phase 10: Cleanup
- Remove legacy Minitest suite (`test_legacy/`) - 187 files deleted
- Update `bin/test` to use RSpec/Cucumber
- Update GitHub Actions workflow with Chrome install and system test support
- Add documentation
- Update README with new test commands and folder structure
- Reorganize `.rubocop.yml` into clear sections
- Create shared context for admin request specs

## Test Counts

| Type | Count | Status |
|------|-------|--------|
| RSpec examples (fast, default) | 678 | ✅ Passing |
| RSpec examples (including slow/system) | 704 | ✅ Passing |
| Cucumber scenarios | 23 | ✅ Passing |
| Cucumber steps | 109 | ✅ Passing |

**Test Reliability**: RSpec suite verified with 5 consecutive passes using random seeds (59498, 19859, 50629, 64555, 12606).

## CI Configuration

GitHub Actions workflow includes:
- Chrome install via `browser-actions/setup-chrome@latest`
- RSpec: `RUN_SLOW_TESTS=true bundle exec rspec`
- Cucumber: `bundle exec cucumber --tags "not @wip"`
- Screenshot upload on test failure (`tmp/capybara/`)

## Files Changed

- **Added**: ~70 new spec/feature files
- **Removed**: 187 legacy test files (~85,000 lines)
- **Modified**: CI workflow, bin/test, seeds, documentation, README

## Documentation

- `doc/adr/0009-test-suite-migration.md` - Migration map, Normal Island diagram, completion status
- `doc/testing-guide.md` - Guide for writing new tests with examples
- `README.md` - Updated testing section and folder structure

## Normal Island Geography

```
NORMAL ISLAND (NO)
├── Northvale (Region)
│   └── Greater Millbrook (County)
│       ├── Millbrook (District)
│       │   ├── Riverside   → NOMB 1RS → Riverside Community Hub
│       │   ├── Oldtown     → NOMB 2OT → Oldtown Library
│       │   ├── Greenfield  → NOMB 3GF → Greenfield Youth Centre
│       │   └── Harbourside → NOMB 4HS → Harbourside Arts Centre
│       └── Ashdale (District)
│           ├── Hillcrest   → NOAD 1HC → Ashdale Sports Club
│           └── Valleyview  → NOAD 2VV
└── Southmere (Region)
    └── Coastshire (County)
        └── Seaview (District)
            ├── Cliffside   → NOSV 1CS → Coastline Wellness Centre
            └── Beachfront  → NOSV 2BF
```

## Test plan

- [x] All RSpec examples pass (704)
- [x] All Cucumber scenarios pass (23)
- [x] Verified 5 consecutive runs with random seeds
- [x] CI workflow runs successfully

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Tickets this resolves

This is a complete rewrite that addresses our current testing issues wholesale, without changing the app functionality itself. It therefore closes the following features and issues arising from our existing tech debt:

Fixes #964
Fixes #1478
Fixes #1919
Fixes #2073
Fixes #2263
Fixes #2460
Fixes #2557
Fixes #2573
Fixes #2591
Fixes #2704

### CODE REVIEW

**Comment by kimadactyl** [comment]
tagging a few people for review - this is obvi a huge change but only touches the test and lint functionality so im not worried about doing so much in one pr, i think this is a pure upgrade from where we are.

my main outstanding q is if we should leave the claude files in the repo or if this is something i should gitignore out?

File: .prettierignore [hunk 1]
```diff
@@ -9,3 +9,4 @@
 /dump/
 /public/assets/
 /coverage/
+/config/cucumber.yml
```
Comment (wheresalice) [inline .prettierignore]: can we instead rename this file to `config/cucumber.yml.erb`?

File: Gemfile [hunk 2]
```diff
@@ -92,12 +91,15 @@ group :development, :test do
...
   gem 'factory_bot_rails'
+  gem 'faker'
+  gem 'rspec-rails', '~> 7.0'
   gem 'timecop'
 end
 
 group :development do
   gem 'better_errors'
   gem 'binding_of_caller'
+  gem 'claude-on-rails'
```
Comment (wheresalice) [inline Gemfile]: what does including the gem `claude-on-rails` do here? What impact does this have on developers who don't want to use AI?

File: README.md [hunk 3]
```diff
@@ -92,16 +92,35 @@ rails events:import_calendar\[100\]
 
 ## Testing, linting and formatting
 
-PlaceCal tests are written in minitest. We use Rubocop to lint our Ruby code.
+PlaceCal uses **RSpec** for unit/integration tests and **Cucumber** for BDD-style acceptance tests. We use Rubocop to lint our Ruby code.
 
 We use [Prettier](https://prettier.io/) to format everything it's able to parse. We run this automatically as part of a pre-commit hook.
 
-You can run the tests & rubocop with:
+### Running tests
```
Comment (wheresalice) [inline README.md]: maybe add a note here about how the database needs to be available to run these tests

File: makefile [hunk 4]
```diff
@@ -47,10 +47,10 @@ seed_GFSC_prod_copy_db:
 	rails db:dump_production_and_restore_other restore_on_local=1
 
 test:
-	time sh -c "rails test --pride && rails test:system && rubocop && prettier -c app/"
+	time sh -c "bundle exec rspec && bundle exec cucumber --tags 'not @wip' && rubocop && prettier -c app/"
 
 tags:
-	find app/ lib/ test/ -iname '*.rb' | xargs etags
+	find app/ lib/ spec/ -iname '*.rb' | xargs etags
```
Comment (wheresalice) [inline makefile]: this needs documentation about needing to install etags as a dependency, though I realise that part isn't changing, so feel free to add as a separate issue

File: README.md [hunk 5]
```diff
@@ -92,16 +92,35 @@ rails events:import_calendar\[100\]
...
+
+# Direct commands
+bundle exec rspec                        # Fast specs only
+RUN_SLOW_TESTS=true bundle exec rspec    # Include system specs
+bundle exec cucumber                     # Cucumber features
 ```
 
+### Test data: Normal Island
+
+Tests use a fictional geography called "Normal Island" (country code: NO) to avoid conflicts with real UK data. See `lib/normal_island.rb` for the full data structure and `doc/testing-guide.md` for guidance on writing tests.
```
[the remark this answers is not in the document]
Comment (kimadactyl) [reply README.md]: Lol ofc. I changed it from NI to not have confusion with Northern Ireland

File: spec/fixtures/files/test_image.jpg
[the remark this answers is not in the document]
Comment (kimadactyl) [reply spec/fixtures/files/test_image.jpg]: i assume its from https://placecats.com/ but ill replace with something from unsplash or the like good call

File: Gemfile (hunk 2 above)
Comment (kimadactyl) [reply Gemfile]: ah - specifically this makes available the `claude-swarm` cli command that loads the prompts. so yeah - i think this is more existential about do we keep this stuff in the codebase or not?

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kimadactyl) [reply README.md]: Renamed to ZZ

File: makefile (hunk 4 above)
Comment (kimadactyl) [reply makefile]: https://github.com/geeksforsocialchange/PlaceCal/issues/2867

File: spec/fixtures/files/test_image.jpg
[the remark this answers is not in the document]
Comment (kimadactyl) [reply spec/fixtures/files/test_image.jpg]: updated with credits

File: Gemfile [hunk 6]
```diff
@@ -92,6 +91,8 @@ group :development, :test do
   gem 'byebug'
   gem 'dotenv-rails'
   gem 'factory_bot_rails'
+  gem 'faker'
+  gem 'rspec-rails', '~> 7.0'
```
Comment (yndajas) [inline Gemfile]: Does this belong in the test (only) group?

File: doc/adr/0009-test-suite-migration.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+- Date: 2025-12-10
+
+## Context and Problem Statement
+
+The current PlaceCal test suite has several issues:
+- 112 test files using Minitest (~11,500 lines, ~500 test methods)
+- Test data uses UK-specific geography (Manchester, Hulme, Moss Side, real postcodes)
+- Tests are coupled to real-world locations which can change
+- No BDD/acceptance test layer exists
+- Factories mix real places with test data, making tests brittle
```
Comment (yndajas) [inline doc/adr/0009-test-suite-migration.md]: Three of these points are pretty similar

File: doc/adr/0009-test-suite-migration.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+│           ├── Ward: Hillcrest
+│           └── Ward: Valleyview
+└── Region: Southmere
+    └── County: Coastshire
+        └── District: Seaview
+            ├── Ward: Cliffside
+            └── Ward: Beachfront
+```
+
+Postcode format: `NO[District] [Ward][Number]` (e.g., `NOMB 1RS` for Riverside)
```
Comment (yndajas) [inline doc/adr/0009-test-suite-migration.md]: `[Number][Ward]`?

File: doc/adr/0009-test-suite-migration.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+2. **Policy specs**: Test all CRUD actions for all roles. Use shared examples.
+3. **Component specs**: Test rendering, slots, and edge cases. Use `render_inline`.
+4. **Request specs**: Test response codes and JSON structure. Don't test UI.
+5. **System specs**: Only for critical user journeys. One flow per test.
+6. **Cucumber**: Business-readable. One scenario per acceptance criterion.
+
+## Implementation Phases
+
+1. Foundation Setup (gems, config, move legacy tests)
+2. Normal Island Data & Factories
```
Comment (yndajas) [inline doc/adr/0009-test-suite-migration.md]: Should the commit message reference phase 2?

File: spec/factories/articles.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    factory :draft_article do
+      is_draft { true }
+      published_at { nil }
+    end
+
+    factory :published_article do
+      is_draft { false }
+      published_at { 1.day.ago }
+    end
```
Comment (yndajas) [inline spec/factories/articles.rb]: Could this just be an alias? `factory :article, aliases: [: published_article]`

File: spec/factories/events.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+    factory :future_event do
+      dtstart { 1.week.from_now.at_beginning_of_hour }
+      dtend { 1.week.from_now.at_beginning_of_hour + 2.hours }
+    end
+
+    factory :today_event do
+      dtstart { Time.current.at_beginning_of_hour + 2.hours }
+      dtend { Time.current.at_beginning_of_hour + 4.hours }
+    end
```
Comment (yndajas) [inline spec/factories/events.rb]: I don't think this would be guaranteed to be today, e.g. if run at 10pm or later

File: spec/factories/partners.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+require_relative '../../lib/normal_island'
+
+FactoryBot.define do
+  factory :partner do
+    sequence(:name) { |n| "Partner #{n}" }
+    summary { Faker::Lorem.paragraph(sentence_count: 2) }
+    description { Faker::Lorem.paragraphs(number: 3).join("\n\n") }
+    public_email { Faker::Internet.email }
+    public_phone { '01onal 234567' }
```
Comment (yndajas) [inline spec/factories/partners.rb]: 👀

**Review by yndajas — COMMENTED** [review]
Only had chance to look at the first two commits so far (so up to phase 2) but in reasonable detail to familiarise myself. No real issues so far. Will continue another day if you still need it!

File: db/seeds/002_neighbourhoods.rb [hunk 13]
```diff
@@ -1,66 +1,85 @@
...
+    # Country
     country = Neighbourhood.create!(
-      name: 'England',
-      name_abbr: '',
-      unit: 'country',
-      unit_code_key: 'CTRY19CD',
-      unit_code_value: 'E92000001',
-      unit_name: 'England',
+      name: NormalIsland::COUNTRY[:name],
+      name_abbr: 'NO',
```
[the remark this answers is not in the document]
Comment (kimadactyl) [reply db/seeds/002_neighbourhoods.rb]: this has been changed to ZZ already :) (see alice's comments)

File: spec/factories/partners.rb (hunk 12 above)
Comment (kimadactyl) [reply spec/factories/partners.rb]: ha honestly i think our phone numbers still allow letters as when we started it phone extensions were still very much a thing - this has already been resolved in the PR but i've opened a ticket for it: 

https://github.com/geeksforsocialchange/PlaceCal/issues/2870
